torch/csrc/jit/codegen/cuda/scheduler/reduction.cpp

```diff
@@ -0,0 +1,543 @@
+#include <torch/csrc/jit/codegen/cuda/scheduler/reduction.h>
+
+#include <torch/csrc/jit/codegen/cuda/executor_utils.h>
+#include <torch/csrc/jit/codegen/cuda/instrumentation.h>
+#include <torch/csrc/jit/codegen/cuda/ir_all_nodes.h>
+#include <torch/csrc/jit/codegen/cuda/scheduler/utils.h>
+
+#include <ATen/cuda/CUDAContext.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+namespace {
+constexpr int kLoopUnrollSplit = 4;
+
+// Largest Power of 2 less-than n
+constexpr int lastPow2(int n) {
```

Comment:
Maybe move to schedule_utils

Reply:
I had it in there but it was complaining about a linking issue I couldn't figure out, so moved it back. Will consider in later refactors to see if I can get it sorted out.